journald_receiver does not work with the filestorage extension

### Component(s)

extension/storage/filestorage, receiver/journald

### What happened?

## Description

journald_receiver does not work as specified with the filestorage extension

## Steps to Reproduce

1. Create an otel collector with the **journaldreceiver** and **filestorageextension**
2. In a config set the journaldreceiver `storage` attribute to the ID of a filestorage component
3. Run the collector with the config

## Expected Result

The collector runs and stores the cursor

## Actual Result

The collector crashes 


### Collector version

v0.95.0

### Environment information

## Environment
OS: Ubuntu 22.04
Compiler: go 1.22

### OpenTelemetry Collector configuration

```yaml
extensions:
  file_storage:
    directory: /var/lib/otelcol/file_storage

receivers:
  journald:
    storage: file_storage
    matches:
      - SYSLOG_FACILITY: "1" # user
    priority: debug

processors:
  batch:
    send_batch_size: 5
    send_batch_max_size: 10
    timeout: 5s

exporters:
  debug:
    verbosity: detailed
    sampling_initial: 5
    sampling_thereafter: 5


service:
  extensions: [file_storage]
  pipelines:
    logs/global:
      receivers: [journald]
      processors: [batch]
      exporters: [debug]
```


### Log output

```shell
24-02-27T12:53:55.896Z        info    adapter/receiver.go:45  Starting stanza receiver        {"kind": "receiver", "name": "journald", "data_type": "logs"}
2024-02-27T12:53:55.896Z        info    service@v0.95.0/service.go:206  Starting shutdown...
2024-02-27T12:53:55.896Z        info    adapter/receiver.go:140 Stopping stanza receiver        {"kind": "receiver", "name": "journald", "data_type": "logs"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x26e4aba]

goroutine 1 [running]:
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/input/journald.(*Input).Stop(0xc000640c60)
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.95.0/operator/input/journald/journald.go:351 +0x1a
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/pipeline.(*DirectedPipeline).stop(0x30?)
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.95.0/pipeline/directed.go:75 +0x17f
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/pipeline.(*DirectedPipeline).Stop.func1()
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.95.0/pipeline/directed.go:48 +0x25
sync.(*Once).doSlow(0x3b5da20?, 0x699cfe0?)
        sync/once.go:74 +0xc2
sync.(*Once).Do(...)
        sync/once.go:65
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/pipeline.(*DirectedPipeline).Stop(0xc000a55d40)
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.95.0/pipeline/directed.go:47 +0x85
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/adapter.(*receiver).Shutdown(0xc00095fef0, {0x4800df0, 0x6a5f2c0})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.95.0/adapter/receiver.go:141 +0x6f
go.opentelemetry.io/collector/service/internal/graph.(*Graph).ShutdownAll(0xc000a058c0, {0x4800df0, 0x6a5f2c0})
        go.opentelemetry.io/collector/service@v0.95.0/internal/graph/graph.go:435 +0x1a8
go.opentelemetry.io/collector/service.(*Service).Shutdown(0xc00095f4d0, {0x4800df0, 0x6a5f2c0})
        go.opentelemetry.io/collector/service@v0.95.0/service.go:212 +0xcf
go.opentelemetry.io/collector/otelcol.(*Collector).setupConfigurationComponents(0xc000979c20, {0x4800df0, 0x6a5f2c0})
        go.opentelemetry.io/collector/otelcol@v0.95.0/collector.go:191 +0x705
go.opentelemetry.io/collector/otelcol.(*Collector).Run(0xc000979c20, {0x4800df0, 0x6a5f2c0})
        go.opentelemetry.io/collector/otelcol@v0.95.0/collector.go:229 +0x52
go.opentelemetry.io/collector/otelcol.NewCommand.func1(0xc000232c08, {0x3ebbfb5?, 0x7?, 0x3eb7179?})
        go.opentelemetry.io/collector/otelcol@v0.95.0/command.go:27 +0x6c
github.com/spf13/cobra.(*Command).execute(0xc000232c08, {0xc000052550, 0x2, 0x2})
        github.com/spf13/cobra@v1.8.0/command.go:983 +0xaca
github.com/spf13/cobra.(*Command).ExecuteC(0xc000232c08)
        github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
github.com/spf13/cobra.(*Command).Execute(0x4066380?)
        github.com/spf13/cobra@v1.8.0/command.go:1039 +0x13
main.runInteractive({0x4066380, {{0x3ee0907, 0x13}, {0x3f27603, 0x24}, {0x3eb80d9, 0x5}}, 0x0, {0x0, 0x0}, ...})
        go.opentelemetry.io/collector/cmd/builder/main.go:27 +0x3d
main.run(...)
        go.opentelemetry.io/collector/cmd/builder/main_others.go:10
main.main()
        go.opentelemetry.io/collector/cmd/builder/main.go:20 +0x118
```


### Additional context

Using the ocb builder to build the collector with the mentioned components. 

The folder /var/lib/otelcol/file_storage is created with the user running the otel collector having access to it.

--- a/.chloggen/journald-protect-shutdown.yaml
+++ b/.chloggen/journald-protect-shutdown.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: journaldreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Fix bug where failed startup could bury error message due to panic during shutdown
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [31476]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []

--- a/pkg/stanza/operator/input/journald/journald.go
+++ b/pkg/stanza/operator/input/journald/journald.go
@@ -348,7 +348,9 @@ func (operator *Input) parseJournalEntry(line []byte) (*entry.Entry, string, err
 
 // Stop will stop generating logs.
 func (operator *Input) Stop() error {
-	operator.cancel()
+	if operator.cancel != nil {
+		operator.cancel()
+	}
 	operator.wg.Wait()
 	return nil
 }

